From: Martin Schwidefsky <schwidefsky@de.ibm.com>
Subject: s390: detect etoken facility
Patch-mainline: v4.19-rc1
Git-commit: aeaf7002a76c8da60c0f503badcbddc07650678c
References: bnc#1106930, LTC#171029

Description:  kernel: improve spectre mitigation
Symptom:      -
Problem:      The upstream kernel has a few more spectre related updates
              and fixes which are relevant for the various distribution
              targets as well. The most notable improvement is the
              etoken support.
Solution:     Backport upstream patches
Reproduction: -

Upstream-Description:

              s390: detect etoken facility

              Detect and report the etoken facility. With spectre_v2=auto or
              CONFIG_EXPOLINE_AUTO=y automatically disable expolines and use
              the full branch prediction mode for the kernel.

              Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>


Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>
Acked-by: Johannes Thumshirn <jthumshirn@suse.de>
---
 arch/s390/kernel/nospec-branch.c |   13 +++++++++++++
 arch/s390/kernel/nospec-sysfs.c  |    2 ++
 2 files changed, 15 insertions(+)

--- a/arch/s390/kernel/nospec-branch.c
+++ b/arch/s390/kernel/nospec-branch.c
@@ -44,6 +44,8 @@ early_param("nogmb", nogmb_setup_early);
 
 static int __init nospec_report(void)
 {
+	if (test_facility(156))
+		pr_info("Spectre V2 mitigation: etokens\n");
 #ifdef CC_USING_EXPOLINE
 	if (!nospec_disable)
 		pr_info("Spectre V2 mitigation: execute trampolines.\n");
@@ -71,6 +73,17 @@ early_param("nospectre_v2", nospectre_v2
 
 void __init nospec_auto_detect(void)
 {
+	if (test_facility(156)) {
+		/*
+		 * The machine supports etokens.
+		 * Disable expolines and disable nobp.
+		 */
+#ifdef CC_USING_EXPOLINE
+			nospec_disable = 1;
+#endif
+		__clear_facility(82, S390_lowcore.alt_stfle_fac_list);
+		return;
+	}
 #ifdef CC_USING_EXPOLINE
 	/*
 	 * The kernel has been compiled with expolines.
--- a/arch/s390/kernel/nospec-sysfs.c
+++ b/arch/s390/kernel/nospec-sysfs.c
@@ -13,6 +13,8 @@ ssize_t cpu_show_spectre_v1(struct devic
 ssize_t cpu_show_spectre_v2(struct device *dev,
 			    struct device_attribute *attr, char *buf)
 {
+	if (__test_facility(156, S390_lowcore.alt_stfle_fac_list))
+		return sprintf(buf, "Mitigation: etokens\n");
 #ifdef CC_USING_EXPOLINE
 	if (!nospec_disable)
 		return sprintf(buf, "Mitigation: execute trampolines\n");
